include/message_filters/sync_policies/latest_time.h

```diff
@@ -0,0 +1,375 @@
+\endverbatim
+ *
+ */
+
+#ifndef MESSAGE_FILTERS__SYNC_LATEST_TIME_H_
+#define MESSAGE_FILTERS__SYNC_LATEST_TIME_H_
+
+#include <algorithm>
+#include <cassert>
+#include <deque>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include <rclcpp/rclcpp.hpp>
+
+#include "message_filters/message_traits.h"
+#include "message_filters/null_types.h"
+#include "message_filters/signal9.h"
+#include "message_filters/synchronizer.h"
+
+
+namespace message_filters
+{
+namespace sync_policies
+{
+
+template<typename M0, typename M1,
+         typename M2 = NullType, typename M3 = NullType, typename M4 = NullType,
+         typename M5 = NullType, typename M6 = NullType, typename M7 = NullType,
+         typename M8 = NullType>
+struct LatestTime : public PolicyBase<M0, M1, M2, M3, M4, M5, M6, M7, M8>
+{
+  typedef Synchronizer<LatestTime> Sync;
+  typedef PolicyBase<M0, M1, M2, M3, M4, M5, M6, M7, M8> Super;
+  typedef typename Super::Messages Messages;
+  typedef typename Super::Signal Signal;
+  typedef typename Super::Events Events;
+  typedef typename Super::RealTypeCount RealTypeCount;
+
+  /// \brief filter coeffs and error margin factor:
+  /// <rate_ema_alpha, error_ema_alpha, rate_step_change_margin_factor>
+  typedef std::tuple<double, double, double> RateConfig;
+
+  LatestTime()
+    : LatestTime(rclcpp::Clock::SharedPtr(new rclcpp::Clock(RCL_ROS_TIME)))
+  {
+  }
+
+  LatestTime(rclcpp::Clock::SharedPtr clock)
+  : parent_(0),
+    ros_clock_{clock}
+  {
+  }
+
+  LatestTime(const LatestTime& e)
+  {
+    *this = e;
+  }
+
+  LatestTime & operator=(const LatestTime & rhs)
+  {
+    parent_ = rhs.parent_;
+    events_ = rhs.events_;
+    rates_ = rhs.rates_;
+    ros_clock_ = rhs.ros_clock_;
+
+    return *this;
+  }
+
+  void initParent(Sync * parent)
+  {
+    parent_ = parent;
+  }
+
+  void setRateConfigPerMessage(const std::vector<RateConfig> & configs)
+  {
+    rate_configs_.assign(configs.begin(), configs.end());
+  }
+
+  void setRateConfig(const RateConfig & config)
+  {
+    rate_configs_.assign(1U, config);
+  }
+
+  template<int i>
+  void add(const typename std::tuple_element<i, Events>::type & evt)
+  {
+    RCUTILS_ASSERT(parent_);
+
+    std::lock_guard<std::mutex> lock(data_mutex_);
+
+    if(!received_msg<i>())
+    {
+      initialize_rate<i>();
+      // wait until we get each message once to publish
+      // then wait until we got each message twice to compute rates
+      // NOTE: this will drop a few messages of the faster topics until
+      //       we get one of the slowest so we can sync
+      std::get<i>(events_) = evt;  // adding here ensures we see even the slowest
+                                   // message twice before computing rate
+      return;
+    }
+
+    std::get<i>(events_) = evt;
+    rclcpp::Time now = ros_clock_->now();
+    std::cout << "compute rate for [" << i << "]" << std::endl;
+    rates_[i].compute_hz(now);
```

Comment:
Please remove this debugging statement.

Reply:
sorry about all the debugs :disappointed: 
removed!